wgengine/tsdns: test gets stuck starting dns ipv6 server

See: https://gitlab.alpinelinux.org/robertgzr/aports/-/jobs/303556

In particular, we are stuck on `tsdns_server_test.go:86`, which is waiting for `NotifyStartedFunc` to be called on the dns.Server.

Based on the already running server in the stacks, I think this the second ipv6 server from `TestDelegate`. But it looks like the tests were run at a commit I don't have, b198a380, so I can't confirm that.

```
panic: test timed out after 10m0s

goroutine 43 [running]:
testing.(*M).startAlarm.func1()
	/usr/lib/go/src/testing/testing.go:1618 +0xe5
created by time.goFunc
	/usr/lib/go/src/time/sleep.go:167 +0x46

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00021f200, 0x5602611def57, 0xc, 0x56026148fb00, 0x560260eb6e01)
	/usr/lib/go/src/testing/testing.go:1169 +0x2da
testing.runTests.func1(0xc000001c80)
	/usr/lib/go/src/testing/testing.go:1439 +0x7a
testing.tRunner(0xc000001c80, 0xc0001dfde0)
	/usr/lib/go/src/testing/testing.go:1123 +0xef
testing.runTests(0xc00000e3e0, 0x5602616c6580, 0xd, 0xd, 0xbffb895e012111a3, 0x8bb2d99b32, 0x5602616cc840, 0x560260e2f510)
	/usr/lib/go/src/testing/testing.go:1437 +0x310
testing.(*M).Run(0xc000210000, 0x0)
	/usr/lib/go/src/testing/testing.go:1345 +0x1eb
main.main()
	_testmain.go:69 +0x138

goroutine 40 [chan receive]:
tailscale.com/wgengine/tsdns.serveDNS(0x5602611dd0c9, 0x7, 0xc0002187e0, 0xc0000de6c0)
	/builds/robertgzr/aports/testing/tailscale/src/src/github.com/tailscale/tailscale/wgengine/tsdns/tsdns_server_test.go:86 +0x13c
tailscale.com/wgengine/tsdns.TestDelegate(0xc00021f200)
	/builds/robertgzr/aports/testing/tailscale/src/src/github.com/tailscale/tailscale/wgengine/tsdns/tsdns_test.go:280 +0x1dd
testing.tRunner(0xc00021f200, 0x56026148fb00)
	/usr/lib/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:1168 +0x2b3

goroutine 41 [IO wait]:
internal/poll.runtime_pollWait(0x7fe4b1a558f8, 0x72, 0x28)
	/usr/lib/go/src/runtime/netpoll.go:222 +0x65
internal/poll.(*pollDesc).wait(0xc000210118, 0x72, 0x200, 0x200, 0xc000117d40)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x47
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).ReadMsg(0xc000210100, 0xc0001cb400, 0x200, 0x200, 0xc000117d40, 0x28, 0x28, 0x0, 0x0, 0x0, ...)
	/usr/lib/go/src/internal/poll/fd_unix.go:242 +0x246
net.(*netFD).readMsg(0xc000210100, 0xc0001cb400, 0x200, 0x200, 0xc000117d40, 0x28, 0x28, 0xc0000c7b68, 0x560260e79934, 0x57bbf52160ea, ...)
	/usr/lib/go/src/net/fd_posix.go:67 +0x92
net.(*UDPConn).readMsg(0xc000010050, 0xc0001cb400, 0x200, 0x200, 0xc000117d40, 0x28, 0x28, 0x5602616ccea0, 0x0, 0x560261424e20, ...)
	/usr/lib/go/src/net/udpsock_posix.go:59 +0x93
net.(*UDPConn).ReadMsgUDP(0xc000010050, 0xc0001cb400, 0x200, 0x200, 0xc000117d40, 0x28, 0x28, 0x560260e9c645, 0x1, 0xc0000c7cf8, ...)
	/usr/lib/go/src/net/udpsock.go:139 +0x9d
github.com/miekg/dns.ReadFromSessionUDP(0xc000010050, 0xc0001cb400, 0x200, 0x200, 0x0, 0x0, 0x5602616cc840, 0xc0001190e0)
	/builds/robertgzr/aports/testing/tailscale/src/pkg/mod/github.com/miekg/dns@v1.1.30/udp.go:42 +0xb9
github.com/miekg/dns.(*Server).readUDP(0xc0002187e0, 0xc000010050, 0x77359400, 0x5602616fc800, 0x560261498d01, 0x5602616fc800, 0x7fe4b1ad0038, 0x5602616fc800, 0x7fe4b1ad0001)
	/builds/robertgzr/aports/testing/tailscale/src/pkg/mod/github.com/miekg/dns@v1.1.30/server.go:645 +0xd9
github.com/miekg/dns.defaultReader.ReadUDP(0xc0002187e0, 0xc000010050, 0x77359400, 0x7fe4b1ad0008, 0xc0001190e0, 0x1, 0x0, 0x560261498e20, 0xc0001190e0)
	/builds/robertgzr/aports/testing/tailscale/src/pkg/mod/github.com/miekg/dns@v1.1.30/server.go:153 +0x45
github.com/miekg/dns.(*Server).serveUDP(0xc0002187e0, 0xc000010050, 0x0, 0x0)
	/builds/robertgzr/aports/testing/tailscale/src/pkg/mod/github.com/miekg/dns@v1.1.30/server.go:462 +0x162
github.com/miekg/dns.(*Server).ListenAndServe(0xc0002187e0, 0x0, 0x0)
	/builds/robertgzr/aports/testing/tailscale/src/pkg/mod/github.com/miekg/dns@v1.1.30/server.go:302 +0x385
tailscale.com/wgengine/tsdns.serveDNS.func2(0xc0000de6c0, 0xc0002187e0)
	/builds/robertgzr/aports/testing/tailscale/src/src/github.com/tailscale/tailscale/wgengine/tsdns/tsdns_server_test.go:82 +0x2d
created by tailscale.com/wgengine/tsdns.serveDNS
	/builds/robertgzr/aports/testing/tailscale/src/src/github.com/tailscale/tailscale/wgengine/tsdns/tsdns_server_test.go:81 +0x125
FAIL	tailscale.com/wgengine/tsdns	600.007s
```

--- a/wgengine/tsdns/tsdns_server_test.go
+++ b/wgengine/tsdns/tsdns_server_test.go
@@ -5,6 +5,9 @@
 package tsdns
 
 import (
+	"log"
+	"testing"
+
 	"github.com/miekg/dns"
 	"inet.af/netaddr"
 )
@@ -71,7 +74,7 @@ func resolveToNXDOMAIN(w dns.ResponseWriter, req *dns.Msg) {
 	w.WriteMsg(m)
 }
 
-func serveDNS(addr string) (*dns.Server, chan error) {
+func serveDNS(tb testing.TB, addr string) (*dns.Server, chan error) {
 	server := &dns.Server{Addr: addr, Net: "udp"}
 
 	waitch := make(chan struct{})
@@ -79,7 +82,11 @@ func serveDNS(addr string) (*dns.Server, chan error) {
 
 	errch := make(chan error, 1)
 	go func() {
-		errch <- server.ListenAndServe()
+		err := server.ListenAndServe()
+		if err != nil {
+			log.Printf("ListenAndServe(%q): %v", addr, err)
+		}
+		errch <- err
 		close(errch)
 	}()
 

--- a/wgengine/tsdns/tsdns_test.go
+++ b/wgengine/tsdns/tsdns_test.go
@@ -274,14 +274,27 @@ func TestResolveReverse(t *testing.T) {
 	}
 }
 
+func ipv6Works() bool {
+	c, err := net.Listen("tcp", "[::1]:0")
+	if err != nil {
+		return false
+	}
+	c.Close()
+	return true
+}
+
 func TestDelegate(t *testing.T) {
 	tstest.ResourceCheck(t)
+
+	if !ipv6Works() {
+		t.Skip("skipping test that requires localhost IPv6")
+	}
 
 	dnsHandleFunc("test.site.", resolveToIP(testipv4, testipv6, "dns.test.site."))
 	dnsHandleFunc("nxdomain.site.", resolveToNXDOMAIN)
 
-	v4server, v4errch := serveDNS("127.0.0.1:0")
-	v6server, v6errch := serveDNS("[::1]:0")
+	v4server, v4errch := serveDNS(t, "127.0.0.1:0")
+	v6server, v6errch := serveDNS(t, "[::1]:0")
 
 	defer func() {
 		if err := <-v4errch; err != nil {
@@ -371,7 +384,7 @@ func TestDelegate(t *testing.T) {
 func TestDelegateCollision(t *testing.T) {
 	dnsHandleFunc("test.site.", resolveToIP(testipv4, testipv6, "dns.test.site."))
 
-	server, errch := serveDNS("127.0.0.1:0")
+	server, errch := serveDNS(t, "127.0.0.1:0")
 	defer func() {
 		if err := <-errch; err != nil {
 			t.Errorf("server error: %v", err)
@@ -473,7 +486,7 @@ func TestConcurrentSetMap(t *testing.T) {
 func TestConcurrentSetUpstreams(t *testing.T) {
 	dnsHandleFunc("test.site.", resolveToIP(testipv4, testipv6, "dns.test.site."))
 
-	server, errch := serveDNS("127.0.0.1:0")
+	server, errch := serveDNS(t, "127.0.0.1:0")
 	defer func() {
 		if err := <-errch; err != nil {
 			t.Errorf("server error: %v", err)
@@ -752,7 +765,7 @@ func TestTrimRDNSBonjourPrefix(t *testing.T) {
 func BenchmarkFull(b *testing.B) {
 	dnsHandleFunc("test.site.", resolveToIP(testipv4, testipv6, "dns.test.site."))
 
-	server, errch := serveDNS("127.0.0.1:0")
+	server, errch := serveDNS(b, "127.0.0.1:0")
 	defer func() {
 		if err := <-errch; err != nil {
 			b.Errorf("server error: %v", err)

